Add proper implementation of waitForNodeToJoin

ravicious opened the pull request:
The mock implementation which simply waits and returns a bogus server is replaced with an implementation that makes a call to `WaitForConnectMyComputerNodeJoin` from #30905. The description of that PR goes into more detail as to what it is for and how to test it.

ravicious committed:
Implement WaitForConnectMyComputerNodeJoin in Electron app

ravicious committed:
wait: Use addEventListener instead of onabort

ravicious committed:
Make TshAbortController emit abort event only once

This aligns it with a regular AbortController, which also emits the event
only once.

ravicious committed:
Refactor how types are imported in tshd fixtures

ravicious commented on web/packages/teleterm/src/ui/ConnectMyComputer/connectMyComputerContext.test.tsx:
```diff
@@ -111,7 +119,6 @@ test('startAgent re-throws errors that are thrown while spawning the process', a
   await act(async () => {
     [, error] = await result.current.startAgent();
   });
-  expect(error).toBeInstanceOf(AgentProcessError);
```
After I added an actual implementation, this test was failing without the mock above. However, since it first checked only the class of the error, the failure message was largely unhelpful.

Instead, we now first check the details behind the error and only then check its class. This way the reported failure shows what the expected error message is vs what was actually returned.

ravicious commented on web/packages/teleterm/src/ui/ConnectMyComputer/connectMyComputerContext.test.tsx:
```diff
@@ -136,7 +149,10 @@ test('starting the agent flips the workspace autoStart flag to true', async () =
-  await act(() => result.current.startAgent());
+  await act(async () => {
+    const [, error] = await result.current.startAgent();
+    expect(error).toBeFalsy();
```
Similar to the comment above, this test was failing without a mock, but the test itself was written in a way where the failure message was not helpful.

Checking the error of `startAgent` both makes the test more correct and also helps with debugging.

ravicious commented on web/packages/teleterm/src/ui/services/connectMyComputer/connectMyComputerService.ts:
```diff
@@ -83,26 +82,13 @@ export class ConnectMyComputerService {
 
   async waitForNodeToJoin(
```
I realized that when the node fails to join the cluster within the timeout, we should include its logs in the output. I'm going to work on implementing this, in the meantime I submitted a PR which already does 90% of the work.

ravicious replied:
This will require being able to get last 10 logs from the agent at any given time. Doable, we'd just need to store the logs in the AgentRunner state.

https://github.com/gravitational/teleport/blob/91847986b249e7bc835b02f28966164954ef5ddb/web/packages/teleterm/src/mainProcess/agentRunner/agentRunner.ts#L33-L39

What's more, while working on this feature I realized that we do not actually save agent logs anywhere. I incorrectly assumed that they're written to disk by default, but they're actually not and are just sent to stderr.

We should be able to create a logger service per agent and then just save it to disk.

ravicious commented on web/packages/teleterm/src/ui/ConnectMyComputer/DocumentConnectMyComputerSetup/DocumentConnectMyComputerSetup.tsx:
```diff
@@ -146,37 +147,40 @@ function AgentSetup({ rootClusterUri }: { rootClusterUri: RootClusterUri }) {
 
   const [createRoleAttempt, runCreateRoleAttempt, setCreateRoleAttempt] =
     useAsync(
-      useCallback(async () => {
```
This version did not await `retryWithRelogin`, so throwing an error here was equal to unhandled error in a promise. Surprisingly though, the attempt status was set properly.

ravicious commented on web/packages/teleterm/src/ui/ConnectMyComputer/connectMyComputerContext.tsx:
```diff
@@ -277,6 +298,7 @@ export const ConnectMyComputerContextProvider: FC<{
         downloadAndStartAgent,
         markAgentAsConfigured,
         isAgentConfiguredAttempt,
+        logsFromStartTimeout,
```
I wasn't able to avoid exporting this, solely due to the fact that in the setup tab we don't have access to `currentAction` and we have to operate entirely on an attempt alone.

https://github.com/gravitational/teleport/blob/01a6f92322a71ae69f4a2db4d4bd18915f950a69/web/packages/teleterm/src/ui/ConnectMyComputer/DocumentConnectMyComputerSetup/DocumentConnectMyComputerSetup.tsx#L237-L244

Long term if we run into a problem like this again, we should just expand `useAsync` so that other than `statusText`, it also includes some kind of an `error` field that holds the whole error. OTOH, an error is not serializable so idk if we can freely put it there.

But ultimately the problem is that `useAsync` is able to represent the error state merely with a string (`statusText`) and that is not enough in some cases. But maybe it's fine to just have "out of band" pieces of state in cases like that. We lose some benefits coming from types, but at least we don't have problems with serializing `useAttempt` attempt object.

avatus replied:
I came across this issue recently with the unified resources call. I wanted to know the specific error being returned from the backend because a 404 or 501 would mean "not implemented" which I could then push around to the legacy view, rather than some actual server issue. I ended up catching it after the call (and had access to the error there) but I agree that it would have been easier if we could just directly access it somewhere in `useAsync` similar to how things like `react-query` work.

gzdunek replied:
AFAIK the state doesn't have to be serializable.

+1 for adding `error` to `useAsync`. I remember that I also had a problem with lack of it.

avatus reviewed (commented):
I love the logs feature. I wasn't able to connect locally and Logs component looked great.
```
2023-08-23T12:05:53-05:00 INFO [PROC:1]    Joining the cluster with a secure token. pid:89508.1 service/connect.go:589
2023-08-23T12:05:53-05:00 INFO [AUTH]      Attempting registration via proxy server. auth/register.go:277
2023-08-23T12:05:53-05:00 INFO [AUTH]      Successfully registered via proxy server. auth/register.go:284
2023-08-23T12:05:53-05:00 INFO [PROC:1]    Instance has obtained credentials to connect to the cluster. pid:89508.1 service/connect.go:646
2023-08-23T12:05:53-05:00 ERRO [PROC:1]    Instance failed to establish connection to cluster: Failed to connect to Proxy Server through tunnel: connection error: desc = "transport: Error while dialing: failed to dial: dial tcp 127.0.0.1:3024: connect: connection refused". pid:89508.1 service/connect.go:123
2023-08-23T12:06:06-05:00 INFO [PROC:1]    Joining the cluster with a secure token. pid:89508.1 service/connect.go:589
2023-08-23T12:06:06-05:00 INFO [AUTH]      Attempting registration via proxy server. auth/register.go:277
2023-08-23T12:06:06-05:00 INFO [AUTH]      Successfully registered via proxy server. auth/register.go:284
2023-08-23T12:06:06-05:00 INFO [PROC:1]    Instance has obtained credentials to connect to the cluster. pid:89508.1 service/connect.go:646
2023-08-23T12:06:06-05:00 ERRO [PROC:1]    Instance failed to establish connection to cluster: Failed to connect to Proxy Server through tunnel: connection error: desc = "transport: Error while dialing: failed
```

Will fix local setup and try again tomorrow

ravicious commented:
The logs you posted seem to exhibit the same issue that I ran into. When the timeout (or any other scenario where we capture the logs) happens, the logs are not properly "flushed" yet, so we end up cutting the last line halfway through.

I'll see if there's an easy way to address this.

ravicious commented:
You might want to try to create a new node outside of Connect My Computer to see if there's something wrong with your cluster config or Connect.

```
teleport node configure --output=$PWD/cluster/config.yaml --proxy=<proxy address> --data-dir=$PWD/cluster/data --token=<static join token>
```

ravicious commented:
Ok, I know what is happening with the cut off log output. strip-ansi-stream in chunking the output too heavily. I'm going to submit a fix in a separate issue.

ravicious commented on web/packages/teleterm/src/ui/ConnectMyComputer/connectMyComputerContext.tsx:
```diff
@@ -71,6 +76,13 @@ export interface ConnectMyComputerContext {
+   *
+   * Prefer to use timeoutLogs from currentAction if possible.
+   */
+  logsFromStartTimeout: string;
```
This would be an awesome idea. I went and implemented it, but unfortunately we need logs from the moment that the timeout has occurred, not from the moment the agent has exited.

Instead of seeing logs related to problems with connecting, you're going to see something like this:

```
ERROR REPORT:
Original Error: *trace.CompareFailedError teleport process has shutdown
Stack Trace:
	github.com/gravitational/teleport/lib/service/service.go:2746 github.com/gravitational/teleport/lib/service.(*TeleportProcess).RegisterWithAuthServer.func1
	github.com/gravitational/teleport/lib/service/supervisor.go:550 github.com/gravitational/teleport/lib/service.(*LocalService).Serve
	github.com/gravitational/teleport/lib/service/supervisor.go:281 github.com/gravitational/teleport/lib/service.(*LocalSupervisor).serve.func1
	runtime/asm_arm64.s:1197 runtime.goexit
User Message: teleport process has shutdown] pid:2731.1 service:register.instance service/supervisor.go:286
2023-09-15T16:33:32+02:00 WARN [PROC:1]    Unable to reuse Instance client for Node. (not available) pid:2731.1 service/connect.go:1066
```

ravicious replied:
One thing I liked about this proposed model though is that it makes sense that logs are included only with objects that describe some kind of an exited state. Generally we want to display logs only when something wrong has happened.

I think it might make me pull the trigger and add `error` to `useAsync`, which is what we're missing here – an ability to pass the logs through `joinClusterAttempt` somehow.

It's too bad that having `error` in `useAsync` won't help us utilize the compiler as much – after all, we'll be able to pass arbitrary errors there, without having some kind of an enum of possible error states. But we cannot really circumvent that as this is how JS works.

gzdunek replied:
>This would be an awesome idea. I went and implemented it, but unfortunately we need logs from the moment that the timeout has occurred, not from the moment the agent has exited.

😞

>I think it might make me pull the trigger and add error to useAsync, which is what we're missing here – an ability to pass the logs through joinClusterAttempt somehow.

Yeah, I think it will be much better than the `logsFromStartTimeout` property.

ravicious replied:
@gzdunek I'm yet to address the rest of your feedback, but please take a look at the useAsync changes in CMC docs.

I'm going to push a separate PR with the useAsync change, but I'm going to leave a commit with these changes on this branch anyway.

gzdunek replied:
The changes to the documents look good to me!

ravicious commented on web/packages/teleterm/src/services/tshd/createAbortController.ts:
```diff
@@ -37,6 +43,11 @@ export default function createAbortController(): TshAbortController {
   return {
     signal,
     abort() {
+      if (hasBeenAborted) {
+        return;
+      }
+
+      hasBeenAborted = true;
```
This alone isn't enough to mirror how AbortController works in browsers. Once `abort()` has been called on the controller, the signal becomes aborted and you cannot reuse it.

```
> var ac = new AbortController()
< undefined

> ac.signal.addEventListener('abort', () => console.log('abort'))
< undefined

> ac.abort()
abort
< undefined

> ac.abort()
< undefined

> ac.signal.addEventListener('abort', () => console.log('abort'))
< undefined

> ac.abort()
< undefined

> ac.abort()
< undefined
```

https://dom.spec.whatwg.org/#abortsignal-signal-abort

I added a test case to make sure we have this covered.

ravicious committed:
createAbortController: Add signal.aborted, use emitter.once

ravicious committed:
Improve wait function based on Deno implementation

https://github.com/denoland/deno_std/blob/72d6e6641e3cd39ae69fba89a78feab354018ef0/async/delay.ts#L39

ravicious committed:
Add a comment about the events package
